### PR TITLE
Add index for classification attribute

### PR DESCRIPTION
**Pull request by akhil1508**
* Resolves: # https://github.com/nextcloud/server/issues/35136

## Summary
- Added a migration to add an index for the "classification" column of the "calendarobjects" table
## TODO

- [ ] ...

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * Auto-generated migration step: Please modify to your needs!
+ */
+class Version1025Date20221114151721 extends SimpleMigrationStep {
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function preSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
```
Comment (kesselb) [inline apps/dav/lib/Migration/Version1025Date20221114151721.php:23]: Please remove the empty method

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+		return $schema;
+	}
+
+	/**
+	 * @param IOutput $output
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 */
+	public function postSchemaChange(IOutput $output, Closure $schemaClosure, array $options): void {
```
Comment (kesselb) [inline apps/dav/lib/Migration/Version1025Date20221114151721.php:47]: Please remove the empty method

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+namespace OCA\DAV\Migration;
+
+use Closure;
+use Doctrine\DBAL\Schema\SchemaException;
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+/**
+ * Auto-generated migration step: Please modify to your needs!
```
Comment (kesselb) [inline apps/dav/lib/Migration/Version1025Date20221114151721.php:14]: Remove this comment.

**Review by kesselb — COMMENTED** [review]
Thank you :+1:

**Review by tcitworld — COMMENTED** [review]
You also need to bump the app version in `apps/dav/appinfo/info.xml` so that the migrations are executed.

**Comment by tcitworld** [comment]
How long did creating the index took on your database? If it's too much we can move it to the dedicated missing indexes command (see for instance #33405)

**Comment by akhil1508** [comment]
@tcitworld It took 14 seconds to create the index on a database with 1 million entries in the `calendarobjects` table

**Comment by akhil1508** [comment]
@tcitworld Bumped from 1.25.0 to 1.25.1

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+	 * @param Closure $schemaClosure The `\Closure` returns a `ISchemaWrapper`
+	 * @param array $options
+	 * @return null|ISchemaWrapper
+	 * @throws SchemaException
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+		$calendarObjectsTable = $schema->getTable('calendarobjects');
+		$calendarObjectsTable->addIndex(['classification'], 'calobj_clssfction_index');
```
Comment (szaimen) [inline apps/dav/lib/Migration/Version1025Date20221114151721.php:26]: can we somehow check if there is really no index yet, I think this is a usual save-guard

Comment (akhil1508) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php]: Fixed.

Comment (szaimen) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php]: thanks!

**Comment by kesselb** [comment]
Hi @akhil1508, would you mind rebasing the pr?

**Comment by akhil1508** [comment]
@kesselb I did rebase by:

- `git remote add upstream https://github.com/nextcloud/server`
- `git fetch upstream`
- ` git rebase upstream/master`
- `git pull origin dav/classification-index` 
- `git push origin dav/classification-index` 

and I see way too many commits above in the changes now :exploding_head:
Did I do something wrong?

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+declare(strict_types=1);
+
```
Comment (nickvergessen) [inline apps/dav/lib/Migration/Version1025Date20221114151721.php:4]: missing copyright header

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+	 * @throws SchemaException
+	 */
+	public function changeSchema(IOutput $output, Closure $schemaClosure, array $options): ?ISchemaWrapper {
+		/** @var ISchemaWrapper $schema */
+		$schema = $schemaClosure();
+		$calendarObjectsTable = $schema->getTable('calendarobjects');
+		if(!$calendarObjectsTable->hasIndex('calobj_clssfction_index')) {
+			$calendarObjectsTable->addIndex(['classification'], 'calobj_clssfction_index');
+		}
+		return $schema;
```
Comment (nickvergessen) [inline apps/dav/lib/Migration/Version1025Date20221114151721.php:29]:
```suggestion
			$calendarObjectsTable->addIndex(['classification'], 'calobj_clssfction_index');
			return $schema;
		}
		return null;
```

**Review by nickvergessen — DISMISSED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by ChristophWurst** [comment]
@akhil1508 could you squash your commits into one please? thx

**Comment by akhil1508** [comment]
> @akhil1508 could you squash your commits into one please? thx

@ChristophWurst Done

**Review by ChristophWurst — APPROVED** [review]

File: apps/dav/lib/Migration/Version1025Date20221114151721.php [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+namespace OCA\DAV\Migration;
+
+use Closure;
+use Doctrine\DBAL\Schema\SchemaException;
+use OCP\DB\ISchemaWrapper;
+use OCP\Migration\IOutput;
+use OCP\Migration\SimpleMigrationStep;
+
+class Version1025Date20221114151721 extends SimpleMigrationStep {
```
[the remark this answers is not in the document]
Comment (akhil1508) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php:35]: Changing 1025 to 1026 everywhere should do it?

[the remark this answers is not in the document]
Comment (tcitworld) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php:35]: @nickvergessen @ChristophWurst is this needed?

[the remark this answers is not in the document]
Comment (ChristophWurst) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php:35]: I think it's needed for correct sorting of migrations

[the remark this answers is not in the document]
Comment (nickvergessen) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php:35]: Yeah, also helps later on when debugging

[the remark this answers is not in the document]
Comment (akhil1508) [reply apps/dav/lib/Migration/Version1025Date20221114151721.php]: Done. Is it okay now?

**Review by tcitworld — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]

**Comment by akhil1508** [comment]
> can you please fix the conflicts so that we can merge this?

@szaimen Done

**Comment by tcitworld** [comment]
Sorry to ask, can you the app version in `apps/dav/appinfo/info.xml` again?

**Comment by szaimen** [comment]
image

**Comment by akhil1508** [comment]
@szaimen @ChristophWurst @tcitworld Anything for me to do here? I've updated the versions to 27.

**Comment by szaimen** [comment]
Screenshot_20230523_230306_Brave

**Comment by akhil1508** [comment]
@szaimen I ran `bash autoloaderchecker.sh` as requested :+1:

**Comment by kesselb** [comment]
Hi, mind to squash the commits as well?

**Comment by akhil1508** [comment]
> Hi, mind to squash the commits as well?

@kesselb I tried with interactive rebase as usual but it's taking all the merge commits into consideration. Let me try again in some other way.

**Comment by akhil1508** [comment]
@all it automatically requested review while I was trying to squash commits.. sorry :(

@kesselb I've tried squashing with different methods and none of them work as there are multiple merge commits from master in this one's history. Is it okay for you if I open PR in a new branch instead?

**Comment by susnux** [comment]
@akhil1508 it should work with:
```
git remote add upstream https://github.com/nextcloud/server.git
git fetch upstream
git rebase -i upstream/master
```

then mark all of you commits as "s" (for squash) and it should work.

**Comment by akhil1508** [comment]
@susnux I did as you suggested and force-pushed, now there are 33 changes instead of 3 changes :thinking:

**Comment by tcitworld** [comment]
@akhil1508 Is the « Allow edits by maintainers » checkbox checked on the sidebar? Seems I can't force push to fix things.

**Comment by akhil1508** [comment]
@tcitworld Seems I cannot do that as this PR is made from a fork under an organization :(

As per https://github.com/orgs/community/discussions/5634

**Comment by tcitworld** [comment]
Replacement PR at #41111
